### PR TITLE
JENKINS-45497# Replace bitbucket api from v1 by v2

### PR DESCRIPTION
**Pull request by vivek**
Replaces bitbucket v1 apis by v2. Also puts pagination support for branches API. See [JENKINS-45497](https://issues.jenkins-ci.org/browse/JENKINS-45497).

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 1]
```diff
@@ -241,17 +246,12 @@ public BitbucketRepository getRepository() throws IOException, InterruptedExcept
         }
     }
 
-    public void deletePullRequestComment(String pullRequestId, String commentId) throws IOException, InterruptedException {
```
Comment (vivek) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Removing it as its not used anywhere in the code. Also searched jenkinsci and its not used by any plugins.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 2]
```diff
@@ -262,28 +262,23 @@ public void postCommitComment(@NonNull String hash, @NonNull String comment) thr
         }
     }
 
-    public void deletePullRequestApproval(String pullRequestId) throws IOException, InterruptedException {
```
Comment (vivek) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Removing it as its not used anywhere in the code. Also searched jenkinsci and its not used by any plugins.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 3]
```diff
@@ -77,25 +65,42 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
 
 public class BitbucketCloudApiClient implements BitbucketApi {
     private static final Logger LOGGER = Logger.getLogger(BitbucketCloudApiClient.class.getName());
-    private static final String V1_API_BASE_URL = "https://api.bitbucket.org/1.0/repositories/";
     private static final String V2_API_BASE_URL = "https://api.bitbucket.org/2.0/repositories/";
     private static final String V2_TEAMS_API_BASE_URL = "https://api.bitbucket.org/2.0/teams/";
     private static final int MAX_PAGES = 100;
```
Comment (vivek) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: @amuniz @stephenc is MAX_PAGES limit intentional? I see this limit for PR, repos, commits etc. Users have complained about such page limit in context of github/blueocean. I think we should remove it but before I do that, I need to confirm. 

I have added pagination for branches and I am not using this limit .

Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Oooh I don't remember seeing that before... sounds like we should strip it out, there should be no upper limit

**Comment by vivek** [comment]
@reviewbybees

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 4]
```diff
@@ -615,16 +606,20 @@ private String getRequest(String path) throws IOException, InterruptedException
...
-        GetMethod httpget = new GetMethod(path);
+        HeadMethod httpHead = new HeadMethod(path);
         try {
-            executeMethod(client, httpget);
-            return httpget.getStatusCode();
+            executeMethod(client, httpHead);
+            while (httpHead.getStatusCode() == API_RATE_LIMIT_CODE) {
+                Thread.sleep(5000);
+                executeMethod(client, httpHead);
+            }
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: :bug: Waiting forever here until the rate limit is ok to process seems wrong to me. I think failing with an exception (so the user can see it in the build/indexing log would be much better).

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java [hunk 5]
```diff
@@ -23,72 +23,64 @@
...
+import com.cloudbees.jenkins.plugins.bitbucket.client.repository.BitbucketCloudRepository;
+import org.codehaus.jackson.annotate.JsonCreator;
+import org.codehaus.jackson.annotate.JsonProperty;
 
-import java.text.ParseException;
-import java.text.SimpleDateFormat;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 
-@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java]: Why is this removed? It will fail the deserilization if some field is not coming in the JSON response. Do we want that? In general I think it is better to do null checks when needed and fail gracefully if needed.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java [hunk 6]
```diff
@@ -48,7 +48,10 @@
     private BitbucketCloudRepositoryOwner owner;
 
     @JsonProperty("updated_on")
-    private String updatedOn;
+    private Date updatedOn;
+
+    @JsonProperty("slug")
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java]: This shouldn't be needed? The field name already follows the name convention.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java [hunk 7]
```diff
@@ -84,10 +87,25 @@ public void setOwner(BitbucketCloudRepositoryOwner owner) {
...
+    public void setUpdatedOn(Date updatedOn) {
         this.updatedOn = updatedOn;
     }
 
+    @SuppressFBWarnings("EI_EXPOSE_REP")
+    public Date getUpdatedOn() {
+        return updatedOn;
+    }
+
+    @JsonProperty("slug")
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java]: Not needed.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java (hunk 6 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java]: Good catch, I forgot to use it. slug is the repo id. Ideally we should be using uuid but then it doesn't work with our case.

See https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D.

We should simply return slug in getRepositoryName() instead of split fullName. 

```
        if (this.fullName != null) {
            return this.fullName.split("/")[1];
        }
```

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java (hunk 5 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java]: `BitbucketCloudApiClient` now uses single instance of ObjectMapper configured to not fail on unknown properties. Using single instance of ObjectMapper is recommended, avoids repeating such annotations everywhere, better for performance  and is thread safe so no issue.

See:
https://github.com/vivek/bitbucket-branch-source-plugin/blob/49e2d0791848ded3f5281279ef73bf7e7ebdf84b/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java#L747

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 4 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Hmm, I agree with you, there should be some timeout (?). However,  this is the pattern among all http outgoing calls in existing code. for example, see: https://github.com/vivek/bitbucket-branch-source-plugin/blob/49e2d0791848ded3f5281279ef73bf7e7ebdf84b/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java#L586.

@stephenc how's GitHub branch source deals with api rate limit?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java (hunk 5 above)
Comment (amuniz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java]: Ok, then all `@JsonIgnoreProperties` should be removed to be consistent.

Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java]: Yes, already done, will push after some testing

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 4 above)
Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: So we should wait for ever, but check and respond to `Thread.interrupt()` and throw an `InterruptedException`

Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Ok, I will adopt same approach here.

Comment (amuniz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: And print some message in the build log? gh-branch-source does that IIRC

Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: @amuniz Not doing logging in this PR, needs further refactoring conflating this PR.

**Comment by vivek** [comment]
@stephenc @amuniz Answered your concerns, PTAL.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 8]
```diff
@@ -293,20 +294,19 @@ public boolean checkPathExists(@NonNull String branchOrHash, @NonNull String pat
     @CheckForNull
     @Override
     public String getDefaultBranch() throws IOException, InterruptedException {
-        String url = V1_API_BASE_URL + this.owner + "/" + this.repositoryName + "/main-branch";
+        String url = V2_API_BASE_URL + this.owner + "/" + this.repositoryName + "/?fields=mainbranch.name";
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: does this work against Hg repos?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 9]
```diff
@@ -551,52 +524,24 @@ private static void setClientProxyParams(String host, HttpClient client) {
...
-                return client.executeMethod(hc, method);
+    private int executeMethod(HttpMethod httpMethod) throws InterruptedException, IOException {
+        int status = client.executeMethod(httpMethod);
+        while (status == API_RATE_LIMIT_CODE) {
+            if (Thread.interrupted()) {
+                throw new InterruptedException();
             }
-        } else {
-            return client.executeMethod(method);
+            LOGGER.fine("Bitbucket Cloud API rate limit reached, sleeping for 5 sec then retry...");
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: leave a `// TODO` comment to the effect that once Bitbucket start reporting the rate limit reset timestamp, we should instead sleep until that timestamp rather than just blindly waiting 5 seconds

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.google.common.collect.ImmutableList;
+import org.codehaus.jackson.annotate.JsonProperty;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+
+/**
+ * @author Vivek Pandey
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java]: `s/@author.*//g`

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
+package com.cloudbees.jenkins.plugins.bitbucket.client;
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java]: <a href="https://imgflip.com/i/1y9ynw"></a>

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
+package com.cloudbees.jenkins.plugins.bitbucket.client;
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java]: <a href="https://imgflip.com/i/1y9ynw"></a>

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.util.List;
+
+/**
+ * Bitbucket API pagination
+ *
+ * @author Vivek Pandey
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java]: `s/@authour.*//g;`

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudCommit.java [hunk 14]
```diff
@@ -23,15 +23,12 @@
  */
 package com.cloudbees.jenkins.plugins.bitbucket.client.branch;
 
+import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketCommit;
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudCommit.java]: imports not in alpha order

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java [hunk 15]
```diff
@@ -23,72 +23,64 @@
  */
 package com.cloudbees.jenkins.plugins.bitbucket.client.branch;
 
-import org.codehaus.jackson.annotate.JsonIgnoreProperties;
-import org.codehaus.jackson.annotate.JsonProperty;
-
 import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketBranch;
+import com.cloudbees.jenkins.plugins.bitbucket.client.repository.BitbucketCloudRepository;
+import org.codehaus.jackson.annotate.JsonCreator;
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java]: imports not in alpha order, you touched the file, you own the violation ;-)

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPullRequestEvent.java [hunk 16]
```diff
@@ -29,10 +29,13 @@
 import com.cloudbees.jenkins.plugins.bitbucket.client.pullrequest.BitbucketPullRequestValue;
 import com.cloudbees.jenkins.plugins.bitbucket.client.repository.BitbucketCloudRepository;
 import com.cloudbees.jenkins.plugins.bitbucket.client.repository.BitbucketCloudRepositoryOwner;
-import org.codehaus.jackson.annotate.JsonIgnoreProperties;
+import org.apache.commons.lang.StringUtils;
 import org.codehaus.jackson.annotate.JsonProperty;
 
-@JsonIgnoreProperties(ignoreUnknown = true)
+import java.text.ParseException;
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPullRequestEvent.java]: imports not in alpha order

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java [hunk 17]
```diff
@@ -24,20 +24,18 @@
 package com.cloudbees.jenkins.plugins.bitbucket.client.repository;
 
 import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketHref;
+import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRepository;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.codehaus.jackson.annotate.JsonIgnore;
+import org.codehaus.jackson.annotate.JsonProperty;
+import org.codehaus.jackson.map.annotate.JsonDeserialize;
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java]: imports not in alpha order

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java [hunk 18]
```diff
@@ -84,10 +82,16 @@ public void setOwner(BitbucketCloudRepositoryOwner owner) {
         this.owner = owner;
     }
 
-    public void setUpdatedOn(String updatedOn) {
+    @SuppressFBWarnings("EI_EXPOSE_REP2")
+    public void setUpdatedOn(Date updatedOn) {
         this.updatedOn = updatedOn;
```
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java]: much better to just clone the date, e.g. `this.updatedOn = updatedOn == null ? null : (Date)updatedOn.clone();` this findbugs report is generally worth paying attention to. Same for the getter

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java (hunk 12 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java]: Dam it, why? Add a license file to the repo and get done with it. Hate seeing these license headers in each source files! We do not do in blueocean (most in team don't do it).

Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java]: +1 License file

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.StringReader;
+import java.text.SimpleDateFormat;
+import java.util.TimeZone;
+
+/**
+ * Jackson based JSON parser
+ *
+ * @author Vivek Pandey
```
[the remark this answers is not in the document]
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java]: Done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
+package com.cloudbees.jenkins.plugins.bitbucket;
```
[the remark this answers is not in the document]
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java]: Done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java (hunk 10 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java]: done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java (hunk 11 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java]: done

Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudPage.java]: done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java (hunk 12 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketPage.java]: Done.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java (hunk 18 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java]: Done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 4 above)
Comment (amuniz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Well, gh-branch-source is not waiting forever, it is just [throwing an exception](https://github.com/jenkinsci/github-branch-source-plugin/blob/ecfe2e072a56fa9b7e30183b21d526034e926fb7/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L980).

Waiting there forever without any user visible effect is not correct IMO. What if there are other processes making requests to BB in parallel so the rate limit is continuously reached? Is this method going to hang forever? Better throw an exception and let it go to the build/indexing log so the user is aware of the problem.

Sorry, I'm keeping my :bug: here.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 8 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Yes it does. I found a bug in my code thats fixed now.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 9 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudCommit.java (hunk 14 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudCommit.java]: done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPullRequestEvent.java (hunk 16 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPullRequestEvent.java]: done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java (hunk 17 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudRepository.java]: done

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java (hunk 15 above)
Comment (vivek) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/branch/BitbucketCloudBranch.java]: done

**Comment by vivek** [comment]
@stephenc I think all review feedback addressed. PTAL.

**Comment by vivek** [comment]
@amuniz PTAL

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 4 above)
Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: @amuniz we are at a disagreement then. IMHO an SCMSource scan should not fail just because of a timeout, it should block until finished - while being responsive to interrupts. If we want to add a timeout to scanning, we can do that at the branch-api level by having a background thread that does the interrupt at the appropriate time.

As such I view this code as correct and I would :bug: if it threw a timeout. We had a lot of issues previously with branches getting lost because of a failure to behave correctly in this regard

**Comment by amuniz** [comment]
@vivek my previous comment was lost in collapsed sections. Why not throwing an exception when the rate limit is reached? (as the gh-branch-source does)

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 4 above)
Comment (amuniz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: > We had a lot of issues previously with branches getting lost because of a failure to behave correctly in this regard

Wasn't it a bug in branch-api? If the retrieve process fails with an exception then the items clean up should not run? Otherwise there is a bug in gh-branch-source (which is throwing an exception).

**Comment by stephenc** [comment]
> Why not throwing an exception when the rate limit is reached? (as the gh-branch-source does)

The gh-branch-source *does not*.

With gh, we know the rate limit before we make the call and we sleep until the rate limit is acceptable for the call we want to make.

There is a difference between bitbucket is off-line and rate limited... with off-line we have no hope of being able to get the answer, with rate-limited we have a hope.

When you are scanning a really big organization, if we throw an error for a rate limit then you will likely *never* be able to get a full scan and hence the orphaned strategies can never be applied, which is a bad thing.

Ideally Bitbucket would provide headers alerting to the rate limit usage and that way we could guard the appropriate calls with sleeps in advance (and log those sleeps to the listener), however bitbucket does not give us that luxury, so instead the only option is to spin-wait.

**Comment by vivek** [comment]
@amuniz  I have made behavior regarding throwing exception is similar to gh-branch-source, see https://github.com/vivek/bitbucket-branch-source-plugin/blob/af080530064379c7bd72ca57d400523fb26a4a3c/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java#L531. See TODO comments couple line below: that is once bitbucket apis start telling us rate limit expiry time, we can optimize this code better.

**Comment by amuniz** [comment]
@stephenc 
> The gh-branch-source does not.

What's [this](https://github.com/jenkinsci/github-branch-source-plugin/blob/ecfe2e072a56fa9b7e30183b21d526034e926fb7/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L980) then?

@vivek 
> I have made behavior regarding throwing exception is similar to gh-branch-source

Unless I'm missing something, it is not the same behaviour (see the link to GH branch source).
The `Thread.interrupt()` is only going to stop the thread if the user choose to cancel the process from the UI.

**Comment by stephenc** [comment]
> What's this then?

Dead code that should have been removed, see [this](https://github.com/jenkinsci/github-branch-source-plugin/blob/ecfe2e072a56fa9b7e30183b21d526034e926fb7/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L826), [this](https://github.com/jenkinsci/github-branch-source-plugin/blob/ecfe2e072a56fa9b7e30183b21d526034e926fb7/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L888), [this](https://github.com/jenkinsci/github-branch-source-plugin/blob/ecfe2e072a56fa9b7e30183b21d526034e926fb7/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L944) and [this](https://github.com/jenkinsci/github-branch-source-plugin/blob/ecfe2e072a56fa9b7e30183b21d526034e926fb7/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L966)

**Comment by amuniz** [comment]
> Dead code

Oh, yeah...........

Anyways, @vivek made me notice that the waiting loop was already there, not added in this PR, so I'm not going to block this because of a previous mistake. Approving.

**Comment by i386** [comment]
@vivek nice job mate :)

**Comment by i386** [comment]
@amuniz @stephenc @vivek can we roll a release for this change?

**Comment by casz** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-47675 I'll try and get to debug it, tomorrow.
